run/idp-sql/main.py

```diff
@@ -0,0 +1,106 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import signal
+import sys
+from types import FrameType
+
+from flask import Flask, render_template, request, Response
+
+import database
+import middleware
+from middleware import jwt_authenticated, logger
+
+app = Flask(__name__, static_folder="static", static_url_path="")
+
+
+@app.before_first_request
+def create_table() -> None:
+    database.create_tables()
+
+
+@app.route("/", methods=["GET"])
+def index() -> str:
+    context = database.get_index_context()
+    cats_count = context["cats_count"]
+    dogs_count = context["dogs_count"]
+
+    lead_team = ""
+    vote_diff = 0
+    leader_message = ""
+    if cats_count != dogs_count:
+        if cats_count > dogs_count:
+            lead_team = "CATS"
+            vote_diff = cats_count - dogs_count
+        else:
+            lead_team = "DOGS"
+            vote_diff = dogs_count - cats_count
+        leader_message = (
+            f"{lead_team} are running by {vote_diff} vote{'s' if vote_diff > 1 else ''}"
+        )
+    else:
+        leader_message = "CATS and DOGS are evenly matched!"
+
+    context["leader_message"] = leader_message
+    context["lead_team"] = lead_team
+    return render_template("index.html", **context)
+
+
+@app.route("/", methods=["POST"])
+@jwt_authenticated
+def save_vote() -> Response:
+    # Get the team and time the vote was cast.
+    team = request.form["team"]
+    uid = request.uid
+    time_cast = datetime.datetime.utcnow()
+    # Verify that the team is one of the allowed options
+    if team != "CATS" and team != "DOGS":
+        logger.warning(f"Invalid team: {team}")
+        return Response(response="Invalid team specified.", status=400)
+
+    try:
+        database.save_vote(team=team, uid=uid, time_cast=time_cast)
+    except Exception as e:
+        # If something goes wrong, handle the error in this section. This might
+        # involve retrying or adjusting parameters depending on the situation.
+        logger.exception(e)
+        return Response(
+            status=500,
+            response="Unable to successfully cast vote! Please check the "
+            "application logs for more details.",
+        )
+
+    return Response(
+        status=200,
+        response="Vote successfully cast for '{}' at time {}!".format(team, time_cast),
+    )
+
+
+# https://cloud.google.com/blog/topics/developers-practitioners/graceful-shutdowns-cloud-run-deep-dive
+def shutdown_handler(signal: int, frame: FrameType) -> None:
+    logger.info("Signal received, safely shutting down.")
+    database.shutdown()
+    middleware.logging_flush()
+    print("Exiting process.", flush=True)
+    sys.exit(0)
```

Comment:
I've heard that this might not be the right way to exit, is there a way to shutdown the flask server?

Reply:
In this case we are using gunicorn to run flask. Gunicorn handles signals to configure graceful timeouts on TERM ([gunicorn docs](https://docs.gunicorn.org/en/stable/signals.html)) However, we default our gunicorn invocations with `--timeout 0` ([gunicorn docs](https://docs.gunicorn.org/en/stable/settings.html?highlight=timeout#timeout)) ([stack overflow](https://stackoverflow.com/questions/60537977/critical-worker-timeout-in-logs-when-running-hello-cloud-run-with-python-f)). There is docs for flask regarding `werkzeug.server.shutdown`, but I believe that is only useful when having request-based shutdown, and I don't believe with gunicorn that werkzerg would be running anyway.

Reply:
sgtm. Thanks for explaining.